### PR TITLE
2.16.0 and Refactor: Prisma Client API reference

### LINKED ISSUE #352: Add `disconnect` option for relations in docs
As per this [message](https://prisma.slack.com/archives/CKQTGR6T0/p1589453192223300) in Slack, it would be better if the `disconnect` option and it's use-case it added to the docs.

slack convo

### LINKED ISSUE #1162: Expand on what [duration] includes in logs
Context: https://prisma-company.slack.com/archives/CEYCG2MCN/p1610348703207600

### LINKED ISSUE #1168: Searching for "lifecycle hooks" should yield a result on the "Middlewares" page
Lots of people think about our "middleware" as "lifecycle hooks" of queries. So we should mention this as a keyword on the middlewares page so it comes up when people search for that.

More context here: https://prisma-company.slack.com/archives/C016KUHB1R6/p1610527039076700

### LINKED ISSUE #1226: Prettier is misconfigured
Dear all,

the prettier in this repository is misconfigured, i.e., it formats the content in `/content` weirdly, sometimes breaking stuff. Also, the `package.json` and `package-lock.json` file are formatted - which is not really required ;)
Formatting the `package.json` file, for example, reformats the regex in the `npm run prettify` command itself, breaking the command.

Also, for example, all indent-whitespaces are removed and so on.. 

You can easily verify it by
- pulling the latest `master` branch
- run `npm run prettify`
- open the `package.json` file

### PR DESCRIPTION
**Pull request by mhwelander**

### CODE REVIEW

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 1]
```diff
@@ -0,0 +1,3253 @@
...
+  }
+}
+```
+
+</CmdResult>
+</CodeWithResult>
+
+### <inlinecode>groupBy</inlinecode>
+
+0
```
Comment (nikolasburk) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]:
```suggestion
```

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 2]
```diff
@@ -0,0 +1,3253 @@
...
+
+#### Examples
+
+TBD.
+
+## Model query options
+
+### <inlinecode>select</inlinecode>
+
+`select` defines which fields are included in the selection set that Prisma Client returns. See: [Select fields and include relations](../../../concepts/components/prisma-client/select-fields) <span class="concept"></span>.
```
Comment (nikolasburk) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]:
```suggestion
`select` defines which fields are included in the object that Prisma Client returns. See: [Select fields and include relations](../../../concepts/components/prisma-client/select-fields) <span class="concept"></span>.
```

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 3]
```diff
@@ -0,0 +1,3253 @@
...
+```
+
+</CmdResult>
+</CodeWithResult>
+
+### <inlinecode>include</inlinecode>
+
+`include` defines which relations are included in the result that the Prisma Client returns. See: [Select fields and include relations](../../../concepts/components/prisma-client/select-fields) <span class="concept"></span>.
+
+#### Examples
```
Comment (nikolasburk) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx:1624]: no `CmdResult`s here? 😢

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 4]
```diff
@@ -0,0 +1,3253 @@
...
+
+See also: [Aggregation, grouping, and summarizing](../../../concepts/components/prisma-client/aggregation-grouping-summarizing#select-distinct) <span class="concept"></span>
+
+#### Reference
+
+```ts
+distinct?: Enumerable<UserDistinctFieldEnum>
+```
+
+```ts
```
Comment (nikolasburk) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]: I noticed that you're not using the `index.d.ts` file annotation here. We should probably be consistent with this across the entire page.

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 5]
```diff
@@ -0,0 +1,3253 @@
...
+  id: 'id'
+  name: 'name'
+  email: 'email'
+  profileViews: 'profileViews'
+  role: 'role'
+  coinflips: 'coinflips'
+}
+```
+
+#### Examples
```
Comment (nikolasburk) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx:2160]: is there a reason why the output in the examples is shown by default but for others it's not?

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 6]
```diff
@@ -0,0 +1,3253 @@
...
+  distinct: ['city'],
+})
+```
+
+</Cmd>
+
+<CmdResult>
+
+```js no-lines
+;[
```
Comment (nikolasburk) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx:2185]:
```suggestion
[
```

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 7]
```diff
@@ -0,0 +1,3253 @@
...
+  distinct: ['city', 'country'],
+})
+```
+
+</Cmd>
+
+<CmdResult>
+
+```js no-lines
+;[
```
Comment (nikolasburk) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx:2218]:
```suggestion
[
```

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 8]
```diff
@@ -0,0 +1,3253 @@
...
+  distinct: ['city', 'country'],
+})
+```
+
+</Cmd>
+
+<CmdResult>
+
+```js
+;[
```
Comment (nikolasburk) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx:2259]:
```suggestion
[
```

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 9]
```diff
@@ -0,0 +1,3253 @@
...
+#### Examples
+
+##### Create a new `User` record with a new `Profile` record
+
+```ts
+const user = await prisma.user.create({
+  data: {
+    email: 'alice@prisma.io',
+    profile: {
+|     create: { bio: 'Hello World' },
```
Comment (nikolasburk) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]: Seems like this needs an extra space

```suggestion
|      create: { bio: 'Hello World' },
```

image

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 10]
```diff
@@ -0,0 +1,3253 @@
...
+```
+
+##### Create a new `Profile` record with a new `User` record
+
+```ts
+const user = await prisma.profile.create({
+  data: {
+    bio: 'Hello World',
+    user: {
+|     create: { email: 'alice@prisma.io' },
```
Comment (nikolasburk) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx:2303]:
```suggestion
|      create: { email: 'alice@prisma.io' },
```

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 11]
```diff
@@ -0,0 +1,3253 @@
...
+```
+
+##### Create a new `User` record with a new `Post` record
+
+```ts
+const user = await prisma.user.create({
+  data: {
+    email: 'alice@prisma.io',
+    posts: {
+|     create: { title: 'Hello World' },
```
Comment (nikolasburk) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]:
```suggestion
|      create: { title: 'Hello World' },
```

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 12]
```diff
@@ -0,0 +1,3253 @@
...
+##### Create a new `User` record with two new `Post` records
+
+Because it's a one-to-many relation, you can also create several `Post` records at once by passing an array to `create`:
+
+```ts
+const user = await prisma.user.create({
+  data: {
+    email: 'alice@prisma.io',
+    posts: {
+|     create: [{ title: 'This is my first post' }, { title: 'Here comes a second post' }],
```
Comment (nikolasburk) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]:
```suggestion
|      create: [{ 
        title: 'This is my first post' 
      }, { 
        title: 'Here comes a second post' 
      }],
```

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 13]
```diff
@@ -0,0 +1,3253 @@
...
+```
+
+##### Update an existing `User` record by creating a new `Profile` record
+
+```ts
+const user = await prisma.user.update({
+  where: { email: 'alice@prisma.io' },
+  data: {
+    profile: {
+|     create: { bio: 'Hello World' },
```
Comment (nikolasburk) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]:
```suggestion
|      create: { bio: 'Hello World' },
```

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 14]
```diff
@@ -0,0 +1,3253 @@
...
+```
+
+To use this feature, enable the `uncheckedScalarInputs` in your schema file as shown:
+
+```prisma
+generator client {
+  provider        = "prisma-client-js"
+  previewFeatures = ["uncheckedScalarInputs"]
+}
+```
```
Comment (nikolasburk) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]: not needed any more as of 2.15 :)

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 15]
```diff
@@ -0,0 +1,3253 @@
...
+|  some?: PostWhereInput | null
+  none?: PostWhereInput | null
+}
+```
+
+#### Examples
+
+##### Retrieve all `User` records where _all_ posts are published and a least one related `Post` mentions `Prisma`
+
+```ts
```
Comment (nikolasburk) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx:3371]: seems like the entire code block is falsely indented

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 16]
```diff
@@ -0,0 +1,3253 @@
...
+  const result = await prisma.user.findMany({
```
Comment (nikolasburk) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]: seems like the entire code block is falsely indented

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 17]
```diff
@@ -0,0 +1,3253 @@
...
+```
+
+## Client methods
+
+Client-level methods are prefixed by `$`.
+
+<!-- prettier-ignore -->
+### <inlinecode>$disconnect()</inlinecode>
+
+The `$disconnect()` method closes the database connections that were established when `$connect` was called and stops the process that was running Prisma's query engine. See [Connection management for an overview of `$connect()` and `$disconnect()`](../../..//concepts/components/prisma-client/connection-management).
```
Comment (nikolasburk) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]: IMO closing the link a bit earlier makes this look cleaner/less noisy

```suggestion
The `$disconnect()` method closes the database connections that were established when `$connect` was called and stops the process that was running Prisma's query engine. See [Connection management](../../..//concepts/components/prisma-client/connection-management) for an overview of `$connect()` and `$disconnect()`.
```

image

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 18]
```diff
@@ -0,0 +1,3253 @@
...
+
+#### Remarks
+
+- `*$disconnect()` returns a `Promise`, so you should call it inside an `async` function with the `await` keyword.
```
Comment (nikolasburk) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]:
```suggestion
- `$disconnect()` returns a `Promise`, so you should call it inside an `async` function with the `await` keyword.
```

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 19]
```diff
@@ -0,0 +1,3253 @@
...
+#### Reference
+
+```ts
+$disconnect(): Promise<void>
+```
+
+<!-- prettier-ignore -->
+### <inlinecode>$connect()</inlinecode>
+
+The `$connect()` method establishes a physical connection to the database via Prisma's query engine. See [Connection management for an overview of `$connect()` and `$disconnect()`](../../..//concepts/components/prisma-client/connection-management).
```
Comment (nikolasburk) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]:
```suggestion
The `$connect()` method establishes a physical connection to the database via Prisma's query engine. See [Connection management](../../..//concepts/components/prisma-client/connection-management) for an overview of `$connect()` and `$disconnect()`.
```

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 20]
```diff
@@ -0,0 +1,3253 @@
...
+
+#### Remarks
+
+- `*$connect()` returns a `Promise`, so you should call it inside an `async` function with the `await` keyword.
```
Comment (nikolasburk) [inline content/400-reference/200-api-reference/050-prisma-client-reference.mdx]:
```suggestion
- `$connect()` returns a `Promise`, so you should call it inside an `async` function with the `await` keyword.
```

File: content/200-concepts/100-components/02-prisma-client/035-select-fields.mdx [hunk 21]
```diff
@@ -0,0 +1,161 @@
+---
+title: 'Select fields'
```
Comment (nikolasburk) [inline content/200-concepts/100-components/02-prisma-client/035-select-fields.mdx:2]: I think this page should highlight a couple of more things:

- examples for `include`
- examples for nested combos of `select` and `include`

File: content/200-concepts/100-components/02-prisma-client/035-select-fields.mdx [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+metaTitle: 'Select fields(Concepts)'
+metaDescription: 'This page explains how to select only a subset of a model''s fields and/or include relations ("eager loading") in a Prisma Client query.'
+tocDepth: 2
+---
+
+<TopBlock>
+
+By default, when a query returns records (as opposed to a count), the result includes the **default selection set**:
+
+- **All** scalar fields defined in the Prisma schema
```
Comment (nikolasburk) [inline content/200-concepts/100-components/02-prisma-client/035-select-fields.mdx]:
```suggestion
- **All** scalar fields defined in the Prisma schema (including enums)
```

File: content/200-concepts/100-components/02-prisma-client/050-filtering-and-sorting.mdx [hunk 23]
```diff
@@ -0,0 +1,182 @@
...
+
+### Sorting FAQs
+ 
+#### Can I sort by the property of a nested `model`?
+
+Follow [issue #249 on GitHub](https://github.com/prisma/prisma-client-js/issues/249).
+
+#### Can I sort by an aggregate value (for example, by user's post count)?
+
+Follow [issue #1 on GitHub](https://github.com/prisma/prisma-client-js/issues/1).
```
Comment (nikolasburk) [inline content/200-concepts/100-components/02-prisma-client/050-filtering-and-sorting.mdx]: seems like this was introduced with `groupBy`? https://github.com/prisma/prisma-client-js/issues/1

**Review by nikolasburk — CHANGES_REQUESTED** [review]
Spectacular work Martina, this is soooo good!

I've left a couple of comments, mostly wrt to consistency and formatting. Sorry that I'm so pedantic with that kind of stuff at times 🙈

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 24]
```diff
@@ -0,0 +1,3253 @@
...
+
+#### Properties
+
+| Example property | Example value                 | Description                                                              |
+| -------------- | ----------------------------- | ------------------------------------------------------------------------ |
+| `db`           | `{ url: 'file:./dev_qa.db }'` | The database [connection string](../database-reference/connection-urls). |
+
+#### Remarks
+
+- You must re-generate the Prisma Client each time you add or rename a datasource. Datasource names are included in the generated client.
```
[the remark this answers is not in the document]
Comment (mhwelander) [reply content/400-reference/200-api-reference/050-prisma-client-reference.mdx]: Good point - I think this is one of the few instances of datasource as opposed to data source. I'll pick data source and update the style guide.

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 25]
```diff
@@ -0,0 +1,3253 @@
...
+| Array of log definitions | `[ { level: "info", emit: "event" }, { level: "warn", emit: "stdout" }]` |             |
+
+##### Log levels
+
+| Name    | Description                                          |
+| ------- | ---------------------------------------------------- |
+| `query` | The query sent to the database, including parameters. |
+| `info`  | Information.                                         |
+| `warn`  | Warnings.                                            |
+| `error` | Errors.                                              |
```
[the remark this answers is not in the document]
Comment (mhwelander) [reply content/400-reference/200-api-reference/050-prisma-client-reference.mdx]: https://github.com/prisma/docs/issues/1245

File: content/400-reference/200-api-reference/050-prisma-client-reference.mdx [hunk 26]
```diff
@@ -0,0 +1,3253 @@
...
+  select: {
+    email: true,
+    name: true,
+  },
+})
+```
+
+</Cmd>
+
+<CmdResult>
```
[the remark this answers is not in the document]
Comment (mhwelander) [reply content/400-reference/200-api-reference/050-prisma-client-reference.mdx:983]: It's just me doing the ones I can for now - have created a clean-up GH issue to go through and make sure all examples have CLI output.

**Pull request merged**